[slider] Add support for custom slider icon

adeelibr opened the pull request:
I have added a new prop called `thumb` to the Slider component, which you can use like this.
```
import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Slider from '@material-ui/lab/Slider';
import DeleteIcon from '@material-ui/icons/Delete';

const styles = {
  root: {
    width: 300,
  },
};

class SimpleSlider extends React.Component {
  state = {
    value: 50,
  };

  handleChange = (event, value) => {
    this.setState({ value });
  };

  render() {
    const { classes } = this.props;
    const { value } = this.state;

    return (
      <div className={classes.root}>
        <Typography id="label">Slider Image</Typography>
        <Slider
          value={value}
          aria-labelledby="label"
          onChange={this.handleChange}
          thumb={<img src="/static/images/cards/live-from-space.jpg" alt="" />}
        />
        <Typography id="label">Slider Simple</Typography>
        <Slider value={value} aria-labelledby="label" onChange={this.handleChange} />
        <Typography id="label">Slider Icon</Typography>
        <Slider
          value={value}
          aria-labelledby="label"
          onChange={this.handleChange}
          thumb={<DeleteIcon style={{ color: 'green' }} />}
        />
      </div>
    );
  }
}

SimpleSlider.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(SimpleSlider);

```

adeelibr committed:
initial commit

adeelibr commented:
![sample-slider](https://user-images.githubusercontent.com/16651811/44003423-f0bb4f9c-9e6b-11e8-9007-2406daf457f6.PNG)

This looks something like this.

ksairamya commented:
We have an urgent requirement for this feature. Please do the needful at the earliest.

eps1lon commented on packages/material-ui-lab/src/Slider/Slider.js:
```diff
@@ -463,7 +485,9 @@ class Slider extends React.Component {
             onFocusVisible={this.handleFocus}
-          />
+          >
+            {Thumbnail()}
```
I would change the `Thumbnail` assignment so that you don't have to use this call syntax here. It's more in line with other components and also removes a second call to the function.

adeelibr replied:
Fixed.

eps1lon commented on packages/material-ui-lab/src/Slider/Slider.js:
```diff
@@ -524,6 +548,7 @@ Slider.propTypes = {
    * @ignore
    */
   theme: PropTypes.object.isRequired,
+  thumb: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
```
Please add a block comment and document this property and use `PropTypes.element`.

adeelibr replied:
Fixed.

eps1lon commented on packages/material-ui-lab/src/Slider/Slider.js:
```diff
@@ -119,6 +119,15 @@ export const styles = theme => {
+    thumbTransparent: {
+      backgroundColor: 'transparent',
+    },
+    thumbIcon: {
```
Please add a block comment to document this `className`.

adeelibr replied:
I have added a small documentation to it.

eps1lon reviewed (changes requested):
When you have added the requested block comments run `yarn docs:api`.

Please add tests and examples to the documentation (see `pages/lab/slider.js`).

adeelibr committed:
Added Documentation, Added prop validation, improved the way on how to use classNames, class overirde issue resolved, changes Thumbnail assignment instead of call, added comments

adeelibr commented on packages/material-ui-lab/src/Slider/Slider.js:
```diff
@@ -434,6 +444,16 @@ class Slider extends React.Component {
     const inlineThumbStyles = { [thumbProperty]: `${percent}%` };
 
+    const Thumbnail = () => {
+      if (React.isValidElement(Thumb)) {
```
Changed prop type to React.element, and removed this extra flag.

adeelibr commented on packages/material-ui-lab/src/Slider/Slider.js:
```diff
@@ -454,7 +474,9 @@ class Slider extends React.Component {
           <ButtonBase
-            className={thumbClasses}
+            className={
+              Thumbnail() ? classNames(thumbClasses, classes.thumbTransparent) : thumbClasses
```
I went with option 1 here, updated thumbClasses.

adeelibr commented on packages/material-ui-lab/src/Slider/Slider.js:
```diff
@@ -434,6 +444,16 @@ class Slider extends React.Component {
+      if (React.isValidElement(Thumb)) {
+        return React.cloneElement(Thumb, {
+          ...Thumb.props,
+          className: classes.thumbIcon,
```
I have made changes now if the user adds classes to the thumb, it won't get over written.

adeelibr commented:
I think I made a mistake, I took a pull master and merged it in my feature-slider-thumbnail branch. And now there is a lot of code in the PR.. what should I do?

Also I have no idea that why in `slider.test.js` this test case is failing `should call handlers' I did nothing with that part of code..

eps1lon commented:
The failing tests should've been resolve with #12535. Try rebasing or merging master again. 

Running`yarn docs:api` should limit the changed files to your feature only.

For general advice about PR workflow I recommend https://gist.github.com/Chaser324/ce0505fbed06b947d962.

adeelibr commented:
@eps1lon I took a pull and merged this branch with master again, on running the tests after that merge all the tests where passing. 

On an unrelated topic regarding this pull request, Is there anything we can do about these tests on CircleCI to improve them, I haven't worked with CircleCI before, is there any way I can help with this.

- ci/circleci: test_browser
- ci/circleci: test_build
- ci/circleci: test_material-ui-x

eps1lon commented:
@adeelibr CircleCI is just broken at the moment (https://discuss.circleci.com/t/saving-cache-stopped-working-warning-skipping-this-step-disabled-in-configuration/24423/2). Nothing you can do about it but patiently wait for a fix.

adeelibr commented:
@eps1lon so does that mean my PR is okay, that it can be merged?

eps1lon commented:
Rebase/merge again to include #12573 and we'll see if everything is ok.

adeelibr commented:
One test is still failing `ci/circleci: test_browser` after the merge including #12573

eps1lon commented:
The branch is just very messy due to all the merge commits. In general if any `Should not have any git stage` fails you should check what command was executed by the CI pipeline before the failed command, run that command locally and commit the changes generated by that task.

In this case `yarn docs:api`.

adeelibr commented:
I have ran `yarn docs:api` updated the master branch and merged it in this feature branch, still failing.

eps1lon commented:
Looks like you ran `docs:api` before the merge. When I checkout this PR and run `docs:api` I get the expected changes that should be committed.

adeelibr commented:
@eps1lon i don't know what's happening this is what I am doing right now

`
git checkout master
git pull upstream master
git checkout feature-branch-name
git merge master
git push origin feature-branch-name
`

eps1lon commented:
You are not running `yarn docs:api` after the merge. This should change many files. All of those changes need to be committed and pushed.

adeelibr commented:
@eps1lon sorry, this is what i was doing
```
git checkout master
git pull upstream master
git checkout feature-branch-name
git merge master
yarn docs:api
git commit -am "some message"
git push origin feature-branch-name
```
what i did was I updated, one of my props definition that I added in this PR ran `yarn docs:api` again, committed the code and pushed it. I hope this works now.

adeelibr commented:
still failed..

eps1lon commented:
I'm pretty sure you have your line endings set to `\r\n` instead of `\n`. Check out `https://help.github.com/articles/dealing-with-line-endings/` and change your setup so that `\n` is saved as the line ending only.

This should probably be mentioned in `Contributing.md`. Asking @oliviertassinari if we should document the behavior, warn if we detect `\r\n` when running `yarn docs:api` or change the regex https://github.com/mui-org/material-ui/blob/823fed03727bb6b52dbfc329472640b8e9da0e21/docs/scripts/buildApi.js#L112-L113

adeelibr commented:
I did this 

```
git config --global core.autocrlf true
```

And ran `yarn docs:api`  but no changes to push.

I then saw the `.gitattributes` file 
https://github.com/mui-org/material-ui/blob/823fed03727bb6b52dbfc329472640b8e9da0e21/.gitattributes#L2

It is already configured for the use case that you have mentioned above, so back at square one I think..

adeelibr commented:
@eps1lon I have changed it from CRLF to LF, changed my git configuration's `git config --global core.autocrlf input` added LF support in my IDE which is VSCODE `"files.eol": "\n",` now when I run `yarn docs:api" it does remove the previous files and when it generates new files they are still CRLF by default.

adeelibr commented:
@oliviertassinari can we somehow change the regex https://github.com/mui-org/material-ui/blob/823fed03727bb6b52dbfc329472640b8e9da0e21/docs/scripts/buildApi.js#L112-L113 so that it is not dependent on the user side.

oliviertassinari commented:
@adeelibr I would try to replace the `\n` with `[\r\n]`.

adeelibr commented:
I have made another PR for this, to resolve this CRLF to LF issue. https://github.com/mui-org/material-ui/pull/12584 I have updated the regex, but it's still giving me the same error. @oliviertassinari

adeelibr commented:
This is what I did

```
git checkout master
git pull upstream master
git checkout feature-branch-name
git merge master
```

Then I made some changes in only the files I wanted to be committed.  Which were;
1- docs/src/pages/lab/slider/CustomIconSlider.js
2- docs/src/pages/lab/slider/slider.md
3- packages/material-ui-lab/src/Slider/Slider.js
4- pages/lab/slider.js

After that
```
git commit -am "some message"
git push --force origin feature-branch-name
```

Isn't this how you wanted me to do it. @oliviertassinari 

But it committed those docs files as well. I am really bad at this.. *sigh*

adeelibr committed:
git reset to working docs
